File: src/main/java/org/luxons/sevenwonders/game/moves/MoveType.java
package org.luxons.sevenwonders.game.moves;

import org.luxons.sevenwonders.game.api.PlayerMove;
import org.luxons.sevenwonders.game.cards.Card;

public enum MoveType {
    PLAY {
        @Override
        public Move resolve(int playerIndex, Card card, PlayerMove move) {
            return new PlayCardMove(playerIndex, card, move);
        }
    },
    UPGRADE_WONDER {
        @Override
        public Move resolve(int playerIndex, Card card, PlayerMove move) {
            return new BuildWonderMove(playerIndex, card, move);
        }
    },
    DISCARD {
        @Override
        public Move resolve(int playerIndex, Card card, PlayerMove move) {
            return new DiscardMove(playerIndex, card, move);
        }
    },
    COPY_GUILD {
        @Override
        public Move resolve(int playerIndex, Card card, PlayerMove move) {
            return new CopyGuildMove(playerIndex, card, move);
        }
    };

    public abstract Move resolve(int playerIndex, Card card, PlayerMove move);
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/moves/MoveType.kt
package org.luxons.sevenwonders.game.moves

import org.luxons.sevenwonders.game.PlayerContext
import org.luxons.sevenwonders.game.api.PlayerMove
import org.luxons.sevenwonders.game.cards.Card

enum class MoveType(private val create: (move: PlayerMove, card: Card, context: PlayerContext) -> Move) {
    PLAY(::PlayCardMove),
    PLAY_FREE(::PlayFreeCardMove),
    UPGRADE_WONDER(::BuildWonderMove),
    DISCARD(::DiscardMove),
    COPY_GUILD(::CopyGuildMove);

    internal fun resolve(move: PlayerMove, card: Card, context: PlayerContext): Move = create(move, card, context)
}
